### ISSUE TITLE
Memory leak on batch writes and gets (levelup+leveldown).

### ISSUE THREAD

**Issue by chjj** [Level/leveldown#262]
I'm retrieving many bitcoin txs from my database in raw binary. I saw very high memory usage after requesting 800+ blocks worth of txs from the db and couldn't figure it out.

I finally tried simply throwing away the value after all the get calls and it still filled up memory something awful (I'm talking 1gb+ when it there should have been <100mb mem usage after the operation).

I tried to replicate my database exactly in an example, but it's not quite the same. Nonetheless, this code below will fill up ~200mb-350mb of memory that never gets freed up.

I get a leak on both batch writes and gets separately as well. When using levelup with leveldown instead of raw leveldown, it tends to end up using 20%-50% more memory.

``` js
var crypto = require('crypto');
var leveldown = require('leveldown');
var db = new leveldown(process.env.HOME + '/leak.db');

db.open({
  createIfMissing: true,
  errorIfExists: false,
  compression: true,
  cacheSize: 1 * 1024 * 1024,
  writeBufferSize: 1 * 1024 * 1024
}, function(err) {
  if (err)
    throw err;

  var ids = [];
  var batch = [];
  for (var i = 0; i < 46045; i++) {
    var data = crypto.randomBytes(478);
    var id = 't/t/' + data.slice(0, 32).toString('hex');
    batch.push({ type: 'put', key: id, value: data });
    ids.push(id);
    id = null;
    data = null;
  }

  db.batch(batch, function(err) {
    if (err)
      throw err;

    var pending = ids.length;
    ids.forEach(function(id) {
      db.get(id, { asBuffer: true }, function(err, data) {
        if (err)
          throw err;
        if (!--pending)
          done();
      });
    });
    ids.length = 0;
    ids = null;

    function done() {
      console.log('done. leaking...');
      setInterval(function() {}, 1000);
    }
  });
  batch = null;
});
```

leveldown: v1.4.4
node: v5.7.0
os: arch
uname: Linux hostname 4.3.3-3-ARCH #1 SMP PREEMPT Wed Jan 20 08:12:23 CET 2016 x86_64 GNU/Linux

**Comment by rvagg** [comment]
leveldb can be a memory hog depending on what you throw at it, it increases over time but eventually stabilises but you never really get it back. It's not a leak per se cause you shouldn't be able to see it keep on going up. If you repeat your tests over and over in the same process on the same db what does the memory profile look like? Stick that whole batch + get section into a loop that repeats after a pause, does it let you keep on going or can you exhaust memory?

**Comment by chjj** [comment]
Thanks for the reply. 

I see. It does use more memory if I run the db.get loop several times, but not as linearly as I'd expect. With my actual data as opposed to the fabricated data above, running it once (reading ~20mb of data from 46k records) results in 760mb of resident mem usage for the node process. Running it 8 times results in 1.2gb of res mem usage for the process. Running it 16 times uses 1.4gb. Running it 32 times results in 1.6gb+.

Oddly, if I do the reads in serial (instead of a parallel loop above), running the loop 16 times results in ~800mb: the same mem usage as running it once. Running them in parallel increases mem usage by 200-400mb each time.

**Comment by rvagg** [comment]
Can you compare RSS to `heapUsed` inside Node when looking at growth? that might be instructive.

**Comment by chjj** [comment]
Hmm, you're right. The leveldb's mem usage is _not_ increasing when I do the loop in parallel. Looks like all the keys/strings in js-land just aren't being cleaned up by v8 before I call memoryUsage().

```
1 iteration (serial):
{ rss: 831283200, heapTotal: 20005920, heapUsed: 10660544 }

2 iterations (serial):
{ rss: 846573568, heapTotal: 28261408, heapUsed: 13287376 }

4 iterations (serial):
{ rss: 841240576, heapTotal: 21496864, heapUsed: 13465504 }

8 iterations (serial):
{ rss: 847409152, heapTotal: 22971424, heapUsed: 12490232 }

16 iterations (serial):
{ rss: 850161664, heapTotal: 25916448, heapUsed: 18683392 }

1 iteration (parallel):
{ rss: 914640896, heapTotal: 56123680, heapUsed: 29412800 }

2 iterations (parallel):
{ rss: 943300608, heapTotal: 74698528, heapUsed: 40429432 }

4 iterations (parallel):
{ rss: 1019408384, heapTotal: 107147552, heapUsed: 73711704 }

8 iterations (parallel):
{ rss: 1159372800, heapTotal: 170538272, heapUsed: 146824056 }

16 iterations (parallel):
{ rss: 1293619200, heapTotal: 297315616, heapUsed: 268675408 }
```

This is still sort of a pain. Is there any way to force leveldb to not eat all my memory?

**Comment by rvagg** [comment]
The heapUsed is a bit of a concern, I wouldn't expect it to grow like that.

What happens if you switch to the native batch functionality:

```
var batch = db.batch()
  for (var i = 0; i < 46045; i++) {
    var data = crypto.randomBytes(478);
    var id = 't/t/' + data.slice(0, 32).toString('hex');
    batch.put(id, data);
}
batch.write(function (err) { ... })
```

In theory that should use less V8 memory because it doesn't have to maintain that big array and instead pushes management of chunks you're writing down into leveldb (this uses the native `Batch` functionality more directly than the classic `db.batch(array)` that you're using).

I'd also suggest removing your `writeBufferSize` adjustment, making it smaller than 4Mb could be having serious performance impacts on your db and could even be contributing to your memory usage here cause it's having to compact the log before it's even ready to create a new table file (which are ~2Mb).

**Comment by chjj** [comment]
Sorry, I should have been clear. What I posted above were the results db.get() calls only. The process did no writes at all. I was testing against data that already exists (my actual data). I can try using native batches for writing the original fabricated test data, but the real memory hog seems to be the reads.

**Comment by rvagg** [comment]
ah, interesting, so some context here that may be important—the async operations all use Node's thread pool, so when you're queueing up 46045 operations you're basically throwing them at the 4 threads that Node has available and letting it schedule them. They don't chunk either so when a read is happening it's occupying that thread for that entire time (which is really quick). So part of your memory spike is coming from queuing up that many operations as well as the closures that handle the responses, which doesn't happen if you do it in serial.

It might be interesting to start node with `UV_THREADPOOL_SIZE` increased to some bigger number than the default `4` and see what effect that has. Maybe `UV_THREADPOOL_SIZE=128` (I think that's the max, you'd have to check the libuv source).

Another thing you can try experimenting with is manual queuing, maybe with `async.queue()` with a set concurrency or even https://github.com/mcollina/throughv to manage it via stream backpressure.

**Comment by chjj** [comment]
Yeah, makes sense that it would be strain on the thread pool. Is there any info on leveldb memory usage? After doing a few googles, I can't seem to find anyone with a similar issue (seeing leveldb allocate 800mb to read 20mb of data). It's very discouraging.

**Comment by rvagg** [comment]
no, I can't think of a good resource on this, _but_ in my experience it should settle down over time, particularly as you have churn and the table files start to sort themselves out a bit more, depending on your app you could even seen it come right back down again over time, so there's an initial peak and then an eventual reduction down to something more reasonable. that initial period, however, is not highly predictable and it's the long-term that matters.

**Comment by chjj** [comment]
I see. I'll keep an eye on my process and see what happens. I'll continue searching for some info for this. I guess adjusting my swapiness will be my solution to avoid frustration for now. Thanks for all the help. Marking as closed now.

**Comment by chjj** [comment]
Reopening. In practice, leveldb doesn't seem to reach any ceiling for memory usage. If I leave my blockchain download running all night it eventually OOMs the process. When the process was killed by the kernel, there was 15mb on the js heap (normal for my process) and 2.5gb+ on the native heap. I've ruled out a Buffer leak in my own code by downloading and parsing the same amount of data, but not retrieving or writing any of the actual data to the db. I'm in the process of trying to isolate this and reproduce it.

**Comment by ralphtheninja** [comment]
@chjj It seems we have a good test case at least.

What tools do we have to our disposal to track this down?

**Comment by chjj** [comment]
@ralphtheninja, I've tried altering a lot of my code to rule things out. I can't seem to find the problem, and I can't reproduce it perfectly because that would mean fabricating a database with records inserted and retrieved in the same volume a blockchain sync would. v8-profiler is obviously useless for this, and valgrind's memleak check didn't seem to find anything obvious in leveldown. I really suspect it's somehow related to the amount of data in the database. Back when I was only using leveldb as an index and flat files for data storage, I didn't have this problem.

I can only reproduce it fully with my module. If you guys want to want to see if you can find anything, it's [this branch](https://github.com/bcoin-org/bcoin/tree/db-tx-pool). `$ node bin/node` will start the sync. Process is killed for me almost always around height 180k, but then again, this chromebook only has 4gb memory. Enabling pruning (which has to do several iterator lookups on every tx when a block comes in and potentially delete some records from the db) makes memory usage higher even quicker: `$ node bin/node --prune`. The relevant levelup code is here: https://github.com/bcoin-org/bcoin/blob/db-tx-pool/lib/bcoin/chaindb.js. I would appreciate if anyone would look at it to tell me if there's something obvious I'm missing (I'm pretty sure I'm freeing up all the iterators and whatnot).

**Comment by chjj** [comment]
If I don't store or retrieve the coins (many thousands to hundreds of thousands of records eventually), memory usage is stable. I guess this must be a relatively "small" leak that most people using leveldown don't notice unless they have a massive amount of records in their db. I'm going to play around with this some more.

@rvagg, any chance the lmdb binding will be updated? I'm very tempted to try it.

**Comment by rvagg** [comment]
I'm happy to provide guidance for anyone that wants to try and upgrade lmdb, I don't even recall what state its in. The NAN 2 API change regex set was built in to https://github.com/nodesource/upgrade-utils if you want to start from an automated replacement (won't do the full job but it's a good start).

**Comment by ralphtheninja** [comment]
I can have a look tomorrow to see what's needed to get it up to par.

**Comment by chjj** [comment]
PR up: https://github.com/rvagg/lmdb/pull/13

**Comment by chjj** [comment]
Update: This really must just be the nature of leveldb (or leveldb itself has a leak in it): I ripped the leveldb dep out of leveldown and replaced it with rocksdb (the leveldown code itself is more or less untouched), and I now have totally reasonable memory usage (66mb instead of 800mb-1gb). So much for leveldb. Closing.
